### PR TITLE
Bind builder attestations to signed-tag source

### LINKED ISSUE #76: Bind builder attestations to the signed-tag source archive
## Current behaviour

`ci/release/validate_builder_attestations.py` verifies that signed builder manifests cover every staged core and PHOTON binary, but it does not bind those manifests to the canonical source archive produced from the signed release tag.

`expected_release_entries()` selects staged artifacts using these prefixes:

- `qbit-<version>-`
- `qbit-photon-<version>-`

The source archive is named `qbit-<version>.tar.gz`, so it is not part of either required set. `manifest_covers_release_entries()` consequently accepts a signed builder manifest when the source archive is missing, renamed, has a different digest, or was produced from another source revision, provided the staged binary entries match.

The publish workflow passes the tag name to the validator, but the validator does not resolve the annotated tag target or reconstruct its canonical source archive.

Affected code:

- `ci/release/validate_builder_attestations.py`
  - `expected_release_entries()`
  - `manifest_covers_release_entries()`
  - `validate_artifact_attestations()`
  - `parse_args()` / `main()`
- `.github/workflows/release-publish.yml`
- `contrib/guix/libexec/build.sh`

## Why this matters

The release process separately verifies:

1. The release tag and signer.
2. Builder signatures and staged binary hashes.

It does not currently prove that the source archive attested by those builders was produced from the signed tag target. This leaves a gap in the intended source-to-binary provenance chain.

This does not indicate that the current published release is mismatched. Reconstructing `qbit-0.1.2-testnet4.tar.gz` from `v0.1.2-testnet4^{}` produces:

```text
a0a596f778f1add8175f40931d274e50857928d4bff09b9ee4f69ef2231b19da
```

That digest is present in both operators' core and PHOTON manifests.

## Steps to reproduce

1. Extend the builder-validator fixtures with a canonical `qbit-<version>.tar.gz` entry.
2. Create otherwise-valid signed manifests that cover every staged binary.
3. Test variants where the source entry is:
   - omitted;
   - renamed;
   - assigned a different digest; or
   - generated from another tag target.
4. Run `ci/release/validate_builder_attestations.py`.

The current validator reaches builder quorum for these variants. As a control, changing a staged binary digest correctly prevents that manifest from counting.

## Expected behaviour

The release validator should reconstruct the canonical source archive from the verified annotated-tag target using the same semantics as the Guix build:

```sh
git archive --prefix="qbit-<version>/" --output="qbit-<version>.tar.gz" <tag-target>
```

Every counted core and PHOTON builder manifest should be required to contain the exact source archive basename and digest.

The source must come from the release tag target, not the tag object or `trusted_release_ref`.

## Acceptance criteria

- [ ] Resolve and verify the annotated release tag target.
- [ ] Reconstruct `qbit-<version>.tar.gz` using the Guix archive prefix semantics.
- [ ] Require its exact basename and digest in every counted core manifest.
- [ ] Require the same entry in every counted PHOTON manifest.
- [ ] Reject missing, renamed, wrong-digest, and wrong-tag source entries.
- [ ] Preserve quorum behavior when an individual manifest fails source binding.
- [ ] Add a positive canonical-source test and negative tests for each failure case.
- [ ] Pass the required source/tag context through the Release Publish workflow.
- [ ] Apply the same invariant to the documented local publication path.
- [ ] Emit the canonical source digest as release evidence.
- [ ] Run a no-publish validation against the immutable `v0.1.2-testnet4` inventory.

Related: #69

### PR DESCRIPTION
**Pull request by kiwidream**
## Summary

- reconstruct the canonical `qbit-<version>.tar.gz` archive from the verified annotated-tag target before evaluating builder quorum
- require that exact source basename and SHA256 in every counted core and PHOTON manifest
- pass the verified tag target and public source checkout through the release workflow and record the source binding in release evidence
- add fail-closed coverage for missing, renamed, wrong-digest, wrong-source, lightweight, indirect, and mismatched tag cases
- document the workflow and local-publication requirements

Fixes #76.

## Testing

- [ ] Built locally.
- [x] Ran focused unit or functional tests for the changed area.
- [x] Ran lint or formatting checks relevant to this change.
- [x] Not run. Reason: A qbit binary build was not run because this changes Python release validation, workflow configuration, tests, and documentation only.

Commands and checks:

- `python3 -m unittest ci.release.test_validate_builder_attestations ci.release.test_validate_key_metadata ci.release.test_validate_release_artifacts ci.release.test_verify_testnet_release_posture` — 131 tests passed, with one expected skip for the absent public local-publish helper
- `python3 contrib/guix/test_build_config.py`
- full lint suite using `ci/lint_imagefile` in Docker
- targeted Docker `py_lint`
- Python bytecode compilation, workflow YAML parsing, and `git diff --check`
- no-publish validation against `v0.1.2-testnet4` and `qbit-guix.sigs@0ee0190073bedf4cf6d3ef60b7531a3551966f26`: core 2/2, PHOTON 2/2, source SHA256 `a0a596f778f1add8175f40931d274e50857928d4bff09b9ee4f69ef2231b19da`

## Target Branch

- [x] This PR targets `main` or a maintainer-requested release branch such as `0.1.x`.

Target: `1.0.0`.

## Risk / Review Notes

- [x] Consensus, script, crypto, wallet, P2P, release, CI, or security-sensitive behavior changed.
- [ ] No consensus, script, crypto, wallet, P2P, release, CI, or security-sensitive behavior changed.

Notes:

- This deliberately makes `--source-root` and `--expected-tag-target` required. Existing local publication callers fail closed until they provide the same verified source context as the workflow.
- Validator code still comes from `trusted_release_ref`; archive bytes come only from the verified release-tag target.
- The source archive remains builder-attestation evidence and is not added to the staged GitHub Release upload set.
- Invalid manifests are rejected individually, preserving existing quorum behavior when enough other valid builders remain.

## Docs / Process Impact

Choose exactly one:

- [x] I updated public docs because this PR changes user-visible behavior, integration guidance, release/process guidance, or expected validation.
- [ ] No public docs update needed. Reason:

## libbitcoinpqc Subtree Checklist (if `src/libbitcoinpqc` changed)

Not applicable; the subtree was not changed.

- [ ] Source commit is reachable from an immutable release tag in `Qbit-Org/qbit-libbitcoinpqc`.
- [ ] qbit imports the tagged upstream tree directly without pruning or a curated subtree branch.
- [ ] Subtree import/update was performed with `contrib/devtools/update-libbitcoinpqc-subtree.sh`.
- [ ] `test/lint/libbitcoinpqc-subtree-check.sh` passes locally.
- [ ] Any default tag change in `contrib/devtools/update-libbitcoinpqc-subtree.sh` is intentional and matches `doc/subtrees/libbitcoinpqc.md`.

---
<a href="https://app.blacksmith.sh/Qbit-Org/codesmith/qbit/pr/84"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1786296679&installation_id=141183937&pr_number=84&repository=Qbit-Org%2Fqbit&return_to=https%3A%2F%2Fgithub.com%2FQbit-Org%2Fqbit%2Fpull%2F84&signature=4064609d8c97032141cabf9b982b4c9384bb4fdbab5e0b287bbeb5d815816b78"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

---

> [!NOTE]
> **High Risk**
> Changes release publication trust: builder quorum and workflow tag verification are security-critical; missing `--source-root`/`--expected-tag-target` fails closed for local callers.
> 
> **Overview**
> Builder attestation quorum now depends on **canonical source binding**, not only staged binary manifests.
> 
> `validate_builder_attestations.py` requires `--source-root` and `--expected-tag-target`, reconstructs `qbit-<version>.tar.gz` from the verified annotated tag’s commit (Guix-style `git archive`), and rejects manifests that omit, rename, or mismatch that archive’s SHA256. Tag handling is stricter: annotated tags must point **directly** at a commit (no lightweight or chained tags).
> 
> The **release-publish** workflow adds a post-checkout step to restore and locally verify the GitHub-verified tag object and target, passes the workspace and `target_commitish` into the builder validator, and records source archive name, digest, and tag target in the job summary.
> 
> Tests cover quorum behavior, GitHub outputs, and fail-closed tag/source edge cases; release and Guix docs describe the new required inputs and evidence fields.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7c2398638f017a44582c093d90efa2bd5b848f44. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: .github/workflows/release-publish.yml [hunk 1]
```diff
@@ -403,6 +409,8 @@ jobs:
           args=(
             --artifacts-dir "${ARTIFACTS_DIR}"
             --tag "${TAG_NAME}"
+            --source-root "${SOURCE_ROOT}"
+            --expected-tag-target "${TARGET_COMMITISH}"
```
[the remark this answers is not in the document]
Comment (kiwidream) [reply .github/workflows/release-publish.yml:435]: Fixed in 1135178b6a. `SOURCE_ROOT` and `TARGET_COMMITISH` now live in the `builders` step environment. I also tightened the workflow contract test to slice the `Validate builder attestations` step and require both variables and both CLI arguments within that step, preventing a step-scope regression. The release-artifact tests, workflow YAML parse, `git diff --check`, and the full Docker lint suite pass.

File: ci/release/validate_builder_attestations.py [hunk 2]
```diff
@@ -406,6 +414,135 @@ def run_command(
...
+    object_type = git_stdout(
+        git,
+        source_root,
+        ["cat-file", "-t", tag_ref],
+        action=f"resolve annotated release tag {tag}",
+    )
+    if object_type != "tag":
+        raise BuilderValidationError(
+            f"Release tag {tag} must be an annotated tag object, got {object_type!r}"
+        )
```
[the remark this answers is not in the document]
Comment (kiwidream) [reply ci/release/validate_builder_attestations.py:464]: Fixed in 7c2398638f. Immediately after `actions/checkout`, the workflow now force-refetches the remote tag ref, resolves both `^{tag}` and `^{commit}`, and compares them to the GitHub-verified tag object SHA and target commit before any release validator runs. The workflow contract test now requires that restore-and-compare step and verifies its ordering before local validation. Release-artifact tests, YAML parsing, `git diff --check`, and the full Docker lint suite pass.
